core/rio/hdt/src/main/java/org/eclipse/rdf4j/rio/hdt/part/HDTPart.java

```diff
@@ -0,0 +1,235 @@
+/*******************************************************************************
+ * Copyright (c) 2020 Eclipse RDF4J contributors.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Distribution License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/org/documents/edl-v10.php.
+ *******************************************************************************/
+package org.eclipse.rdf4j.rio.hdt.part;
+
+import java.io.IOException;
+import java.io.InputStream;
+
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.zip.CheckedInputStream;
+
+/**
+ * Helper class for different HDT parts.
+ * 
+ * Each part starts with <code>$HDT</code>, followed by a byte indicating the type of the part.
+ * 
+ * Structure:
+ * 
+ * <pre>
+ * +------+------+
+ * | $HDT | type |
+ * +------+------+
+ * </pre>
+ * 
+ * @author Bart Hanssens
+ */
+public abstract class HDTPart {
+	public enum Type {
+		GLOBAL((byte) 1),
+		HEADER((byte) 2),
+		DICTIONARY((byte) 3),
+		TRIPLES((byte) 4);
+		private final byte value;
+
+		/**
+		 * Get value associated with this type
+		 * 
+		 * @return value 1,2 or 3
+		 */
+		public byte getValue() {
+			return value;
+		}
+
+		private Type(byte value) {
+			this.value = value;
+		}
+	}
+
+	public final static byte[] COOKIE = "$HDT".getBytes();
+
+	protected Map<String, String> properties;
+
+	// TODO: make configurable, buffer for reading object values
+	private final static int BUFLEN = 1 * 1024 * 1024;
+
+	/**
+	 * Parse from input stream
+	 * 
+	 * @param is
+	 * @throws IOException
+	 */
+	public abstract void parse(InputStream is) throws IOException;
+
+	/**
+	 * Get properties, if any.
+	 * 
+	 * @return key,value map
+	 */
+	public Map<String, String> getProperties() {
+		return properties;
+	}
+
+	/**
+	 * Check start of part for <code>$HDT</code> and the byte indicating the type
+	 * 
+	 * @param is    input stream
+	 * @param ctype control type
+	 * @throws IOException
+	 */
+	protected static void checkControl(InputStream is, HDTPart.Type ctype) throws IOException {
+		byte[] cookie = new byte[COOKIE.length];
+		is.read(cookie);
+		if (!Arrays.equals(cookie, COOKIE)) {
+			throw new IOException("$HDT marker not found");
+		}
+
+		byte b = (byte) is.read();
+		if (b != ctype.getValue()) {
+			throw new IOException("Expected different control info type");
+		}
+	}
+
+	/**
+	 * Check for <code>null</code> terminated format string.
+	 * 
+	 * @param is
+	 * @param format
+	 * @throws IOException
+	 */
+	protected static void checkFormat(InputStream is, byte[] format) throws IOException {
+		byte[] b = new byte[format.length];
+		is.read(b);
+		if (!Arrays.equals(b, format)) {
+			throw new IOException("Unknown format, expected " + new String(format));
```

Comment:
The `String(bytes, Charset)` constructor should be used here to specify which Charset to use when interpreting the bytes.

Reply:
done